pkg/profiles/profiles.go

```diff
@@ -306,6 +315,90 @@ func ToRequestMatch(reqMatch *sp.RequestMatch) (*pb.RequestMatch, error) {
 	}, nil
 }
 
+// ValidateSP takes a v1alpha1.ServiceProfile object, marshals to bytes, and
+// wraps `Validate`.
+func ValidateSP(serviceProfile sp.ServiceProfile) error {
+	data, err := yaml.Marshal(serviceProfile)
+	if err != nil {
+		return fmt.Errorf("failed to marshall ServiceProfile [%+v]: %s", serviceProfile, err)
+	}
+	return Validate(data)
+}
+
+// Validate validates the structure of a ServiceProfile. This code is a superset
+// of the validation provided by the `openAPIV3Schema`, defined in the
+// ServiceProfile CRD.
+// openAPIV3Schema validates:
+// - types of non-recursive fields
+// - presence of required fields
+// This function validates:
+// - types of all fields
+// - presence of required fields
+// - presence of unknown fields
+// - recursive fields
+func Validate(data []byte) error {
+	var serviceProfile sp.ServiceProfile
+	err := yaml.UnmarshalStrict([]byte(data), &serviceProfile)
+	if err != nil {
+		return fmt.Errorf("failed to validate ServiceProfile: %s", err)
+	}
+
+	_, _, err = ValidateName(serviceProfile.Name)
+	if err != nil {
+		return err
+	}
+
+	if len(serviceProfile.Spec.Routes) == 0 {
+		return fmt.Errorf("ServiceProfile \"%s\" has no routes", serviceProfile.Name)
+	}
+
+	for _, route := range serviceProfile.Spec.Routes {
+		if route.Name == "" {
+			return fmt.Errorf("ServiceProfile \"%s\" has a route with no name", serviceProfile.Name)
+		}
+		if route.Timeout != "" {
+			_, err := time.ParseDuration(route.Timeout)
+			if err != nil {
+				return fmt.Errorf("ServiceProfile \"%s\" has a route with an invalid timeout: %s", serviceProfile.Name, err)
+			}
+		}
+		if route.Condition == nil {
+			return fmt.Errorf("ServiceProfile \"%s\" has a route with no condition", serviceProfile.Name)
+		}
+		err := ValidateRequestMatch(route.Condition)
+		if err != nil {
+			return fmt.Errorf("ServiceProfile \"%s\" has a route with an invalid condition: %s", serviceProfile.Name, err)
+		}
+		for _, rc := range route.ResponseClasses {
+			if rc.Condition == nil {
+				return fmt.Errorf("ServiceProfile \"%s\" has a response class with no condition", serviceProfile.Name)
+			}
+			err = ValidateResponseMatch(rc.Condition)
+			if err != nil {
+				return fmt.Errorf("ServiceProfile \"%s\" has a response class with an invalid condition: %s", serviceProfile.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// ValidateName validates that a ServiceProfile's name is of the form:
+// <service>.<namespace>.svc.cluster.local
```

Comment:
We can do this in a follow-up but this validation will have to be loosened to support external names.  We should change this to only validate that the name is a valid DNS name.

Reply:
Filed #2220 to track.